### PR TITLE
<fix> refactor the code to separate modules

### PR DESCRIPTION
**Pull request by tfdavids-zz**
@jobevers mind taking a look?

### CODE REVIEW

File: gae_discourse_client/client.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        else:
+            url = '%s%s' % (self._discourse_url, req_string)
+
+        response = yield ndb.get_context().urlfetch(
+            url=url, payload=urlencode(payload), method=method,
+            headers={'Content-Type': 'application/x-www-form-urlencoded'}
+        )
+
+        if response.status_code != 200:
+            raise Error(
```
Comment (jobevers) [inline gae_discourse_client/client.py]: I don't think this Error is defined.

File: gae_discourse_client/client.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+from google.appengine.api import urlfetch
+from google.appengine.ext import ndb
+
+import categories as categories_module
+import groups as groups_module
+import users as users_module
+
+
+class DiscourseAPIClient(object):
```
Comment (jobevers) [inline gae_discourse_client/client.py]: Can you add tests for this class too?

File: gae_discourse_client/client.py (hunk 1 above)
Comment (tfdavids-zz) [inline gae_discourse_client/client.py]: Fixed

File: gae_discourse_client/client.py (hunk 2 above)
Comment (tfdavids-zz) [inline gae_discourse_client/client.py]: I'm not sure how. If our requests to Discourse are broken, all the other tests will fail anyway, and I'm not aware of any generic HTTP requests (`GET`s or `POST`s or `DELETE`s) that I could make. Did you have any specific tests in mind that you wanted to see?

**Comment by ods94065** [comment]
I think the test setup here is still too non-standard. Here's what I'd like to see instead:

Either:
- a `tests/` directory with two subdirectories, `unit/` and `functional/` (option: you can replace "functional" with "end-to-end" or "acceptance" if you prefer)
- a `tests/` directory with just the unit tests, and a separate `functional_tests/` directory (with vocab note above)

Then:
- configure `python setup.py test` to run just the unit tests, by following directions at http://nose2.readthedocs.org/en/latest/usage.html to modify the setup script, and making appropriate changes to nose2.cfg and maybe setup.cfg if necessary.
- Optionally add a command `test_functional` to setup.py that runs the functional tests, and/or a command `test_all` that tests all of them, following the documentation at https://docs.python.org/2/distutils/extending.html .

I think all this is reasonably in keeping with Python standards, though I admit I don't know of a lot of Python clients like this with functional tests baked in, so we might be doing some trail-hacking here.

Thoughts, @moredatarequired ?

**Comment by moredatarequired** [comment]
Hey! I've just been working through Test Driven Development with Python, which is actually a pretty good book and a nice way to learn Django and some front-end stuff as well (not that I expect to need Django for anything). Despite having done TDD with Python before, and being kind of a Zealot about it on occasion, I still feel like it's a good coverage of standard practice in the Python community.

Anyway, I don't have anything to add to @ods94065's comments. Using a src/ directory or a double-named (gae-discourse-client/gae-discourse-client/) seems pretty common, and I might push all your non-test code into src/, but the test structure & runner he suggests sounds good.

File: gae_discourse_client/categories.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+class CategoryClient(object):
+    """An API client for interacting with Discourse for category-related actions"""
+
+    def __init__(self, api_client):
+        self._api_client = api_client
+
+    @ndb.tasklet
+    def getByName(self, category_name, parent_category_name=None):
```
Comment (ods94065) [inline gae_discourse_client/categories.py:22]: General thought: for libraries that are being published, it's a good idea to spend more time on the docs than you normally would. I'd like to see docstrings for all the functions in these client classes, particularly documenting the args and return values. Use the Google formatting style, described [here](http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_google.html).

The ultimate goal for a Python library would be to get Python documentation compiled with Sphinx and uploaded to http://readthedocs.org, as described [here](http://python-guide.readthedocs.org/en/latest/writing/documentation/), but I think all that's overkill for now unless you really want to learn how to do it.

File: gae_discourse_client/discourse_client.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
+"""Gateway for accessing the Discourse API (for forums)"""
+
+import json
+import re
+from urllib import urlencode
```
Comment (ods94065) [inline gae_discourse_client/discourse_client.py]: Style nit: don't import bare functions into scope; simply `import urllib` instead.

File: gae_discourse_client/groups.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
+"""Gateway for accessing the Discourse API (for forums)"""
+
+import json
+import re
+from urllib import urlencode
```
Comment (ods94065) [inline gae_discourse_client/groups.py]: See above. Though is `urlencode` necessary in this module?

File: setup.py [hunk 6]
```diff
@@ -21,8 +22,14 @@ def read(fname):
     license="MIT",
     keywords="google appengine discourse api client",
     url="https://github.com/udacity/gae-discourse-client",
-    packages=['src', 'tests'],
-    long_description=read('README'),
+    packages=['gae_discourse_client', 'tests', 'functional_tests'],
+    test_suite='nose2.collector.collector',
+    entry_points={
+        "distutils.commands": [
+            "test_functional = test_functional:FunctionalTests"
```
Comment (ods94065) [inline setup.py:29]: We should make sure the `README.md` mentions how to run unit and functional tests through setup.py, including environment setup. Also, if the user has to install requirements.txt before running tests, that should be mentioned too.

**Comment by ods94065** [comment]
Those are the only bits I have left! Other than that, LGTM. Be sure to squash before merging.

**Pull request merged**